db_stress_tool/db_stress_test_base.cc

```diff
@@ -2883,6 +2883,10 @@ void InitializeOptionsFromFlags(
   options.blob_compaction_readahead_size = FLAGS_blob_compaction_readahead_size;
   options.blob_file_starting_level = FLAGS_blob_file_starting_level;
 
+  if (FLAGS_cache_size > 0) {
```

Comment:
I wonder whether we should use a separate flag for blob cache so that the feature can be enabled/disabled independent of block cache by modifying db_crashtest.py

Reply:
I have no problem with either option. But generally speaking, in the production environment, when you turn on blob cache, you will also turn on block cache. This is because you need to cache blob pointers in the block cache. @ltamasi

Reply:
+1 for @riversand963 's suggestion. I agree that having a *blob* cache but no *block* cache is not a realistic scenario; however, I can see a case for having a *block* cache but no *blob* cache in e.g. low-memory environments.

Reply:
One more thing related to this - we would want to support putting blocks and blobs in the same cache, as well as putting them in their own separate caches. This could be an additional flag (something like "use shared block and blob cache").

Reply:
> when you turn on blob cache, you will also turn on block cache

Functionality-wise, I think application today can disable block cache. Do you mean, if the user disables block cache, blob cache is not usable?

Reply:
Taking a look at options, I think the application should be able to do "no block cache" but blob cache. Let me know if I miss anything.

Reply:
> Taking a look at options, I think the application should be able to do "no block cache" but blob cache. Let me know if I miss anything.

Right, it is definitely possible to configure it that way; we just don't expect people to use it like that since SST blocks are generally considered more valuable targets for caching.

Anyways, I think what we would want in `db_stress` and `db_bench` is two things: 1) a way to configure a separate blob cache and 2) a command-line flag to enable using the block cache for blobs as well.

I think it might make sense to bail out with an error if both 1 and 2 are specified on the command line, or if 2 is specified but there is no block cache configured.

Reply:
> > Taking a look at options, I think the application should be able to do "no block cache" but blob cache. Let me know if I miss anything.
> 
> Right, it is definitely possible to configure it that way; we just don't expect people to use it like that since SST blocks are generally considered more valuable targets for caching.
> 
> Anyways, I think what we would want in `db_stress` and `db_bench` is two things: 1) a way to configure a separate blob cache and 2) a command-line flag to enable using the block cache for blobs as well.
> 
> I think it might make sense to bail out with an error if both 1 and 2 are specified on the command line, or if 2 is specified but there is no block cache configured.

If the user specified `read_options.read_tier = kBlockCacheTier;` and we have a blob cache but no block cache, what will happen?  I'll keep an eye on these corner cases...

Reply:
Thanks for the clarification

Reply:
> If the user specified `read_options.read_tier = kBlockCacheTier;` and we have a blob cache but no block cache, what will happen?  I'll keep an eye on these corner cases...

All their reads would return `Status::Incomplete` since they would require I/O which is not allowed. Note that the presence or absence of a blob cache is irrelevant here; using `kBlockCacheTier` but disabling the block cache is enough to hit this scenario (which BTW is clearly a user/application error).

Reply:
How about those three flags?

```c++
DEFINE_bool(use_blob_cache, false, "[Integrated BlobDB] Enable blob cache.");

DEFINE_bool(use_shared_block_and_blob_cache, true,
            "[Integrated BlobDB] Use a shared backing cache for both block "
            "cache and blob cache.");

DEFINE_int64(
    blob_cache_size, FLAGS_cache_size,
    "[Integrated BlobDB] Number of bytes to use as a cache of blobs. It only "
    "takes effect if the block cache and blob cache are different "
    "(use_shared_block_and_blob_cache = false).");
```

Reply:
Since we still use block cache as the global memory limit if the write buffer manager is enabled, blob_cache_size should be smaller than the general cache size for block cache. I guess we don't have to worry about this for now.